### PR TITLE
SPIRVProducerPass refactor - phase2.1

### PR DESCRIPTION
**Pull request by sjw36**
* Removed pointer management, SPIRVInstructionLists are static
* Changed DeferredInst mechanism, addSPIRVPlaceholder then replace in 2nd pass (HandleDeferredInstruction)

Enhancement for clspv issue #545

### CODE REVIEW

File: lib/SPIRVProducerPass.cpp [hunk 1]
```diff
@@ -157,31 +160,26 @@ struct SPIRVOperand {
 private:
   SPIRVOperandType Type;
   std::string LiteralStr;
-  SmallVector<uint32_t, 4> LiteralNum;
+  uint32_t LiteralNum[2];
```
Comment (sjw36) [inline lib/SPIRVProducerPass.cpp:163]: Simplified to make copy light-weight, so these can be collected by value vs pointer management.

**Comment by sjw36** [comment]
This doesn't look like a formatting error:

  + tar xf clang-llvm.tar.xz
  tar: This does not look like a tar archive
  xz: (stdin): File format not recognized
  tar: Child returned status 1
  tar: Error is not recoverable: exiting now

**Comment by alan-baker** [comment]
Yes, LLVM server has been returning 503s lately.

**Comment by alan-baker** [comment]
Can you rebase on master? I've update some bot scripts.

**Comment by sjw36** [comment]
@alan-baker should be up-to-date

**Comment by sjw36** [comment]
Hmm, looks like my local clang-formatter is not right anymore..

**Comment by sjw36** [comment]
Now it's pointing at the right clang-formatter, I think.

File: lib/SPIRVProducerPass.cpp [hunk 2]
```diff
@@ -222,30 +218,29 @@ struct SPIRVInstruction {
...
 struct SPIRVProducerPass final : public ModulePass {
   typedef DenseMap<Type *, uint32_t> TypeMapType;
   typedef UniqueVector<Type *> TypeList;
-  typedef DenseMap<Value *, uint32_t> ValueMapType;
+  typedef DenseMap<Value *, SPIRVID> ValueMapType;
   typedef UniqueVector<Value *> ValueList;
   typedef std::vector<std::pair<Value *, uint32_t>> EntryPointVecType;
-  typedef std::list<SPIRVInstruction *> SPIRVInstructionList;
+  typedef std::list<SPIRVInstruction> SPIRVInstructionList;
   // A vector of tuples, each of which is:
```
Comment (alan-baker) [inline lib/SPIRVProducerPass.cpp:239]: This comment is no longer correct.

File: lib/SPIRVProducerPass.cpp [hunk 3]
```diff
@@ -4856,18 +4862,19 @@ bool SPIRVProducerPass::is4xi8vec(Type *Ty) const {
...
-      while ((*InsertPoint)->getOpcode() == spv::OpPhi) {
-        ++InsertPoint;
-      }
-    }
+  for (size_t i = 0; i < DeferredInsts.size(); ++i) {
+    Value *Inst = DeferredInsts[i].first;
+    SPIRVInstruction *Placeholder = DeferredInsts[i].second;
+    SPIRVOperandVec Operands;
+
+    auto nextDeferred = [&]() {
```
Comment (alan-baker) [inline lib/SPIRVProducerPass.cpp:4872]: No capture alls please. Only capture what is required.

**Pull request merged**